r/R/dotplot.R

```diff
@@ -1,35 +1,84 @@
+#'               useMarkerGenes: True/False determines whether to use marker genes or predefined genes
+#'               numberOfMarkers: Int. Number of marker genes to use
+#'               customGenesList: List of Strings. List of marker genes to use
+#'               groupBy: Cellsets to show in the plot. Determines whether to show Louvain/Samples/Custom
+#'               filterBy: Cellsets to subset the experiment with.
+#'               isFilterByAll: Bool value. If true, the experiment will be subsetted to all the cellSets ids.
+#'              }
+#'            }
+#' @param data
+#'
+#' @return
+#' @export
+#'
+#' @examples
 runDotPlot <- function(req, data) {
-  type <- req$body$type
-  req_genes <- req$body$genes
+  useMarkerGenes <- req$body$useMarkerGenes
+  data$dotplot.groups <- NA
+  group_by_cell_sets <- req$body$groupBy$children
+  filter_by <- req$body$filterBy
+  apply_filter <- req$body$applyFilter
 
-  # construct clusters from cell sets
-  data$custom <- NA
-  cells_id <- data$cells_id
-  cell_sets <- req$body$cellSets$children
+  if (length(group_by_cell_sets) < 1) {
+    message("The requested Cell Sets are empty. Returning empty results.")
+    return(list())
+  }
 
-  for (i in seq_along(cell_sets)) {
-    set <- cell_sets[[i]]
-    filtered_cells <- intersect(set$cellIds, cells_id)
-    data$custom[cells_id %in% filtered_cells] <- i
+  #Collect ids to subset object
+  if (apply_filter) {
+    subsetIds <- filter_by$cellIds
+  } else {
+    subsetIds <- list()
+    for (child in filter_by$children) {
+      subsetIds <- append(subsetIds, child$cellIds)
+    }
   }
 
-  if (type == "custom") {
-    annot <- data@misc$gene_annotations
-    annot_subset <- subset(annot, toupper(name) %in% toupper(req_genes))
-    annot_subset <- annot_subset[, c("input", "name")]
+  #Subset cell Ids
+  if (length(subsetIds)) {
+    data <- subset_ids(data,subsetIds)
+    cells_id <- data$cells_id
   } else {
-    # type == marker
-    # Need the marker features
+    message("The ids to subset the object are empty. Returning empty results.")
+    return(list())
   }
 
-  dotplot_data <- Seurat::DotPlot(data, features = annot_subset$input, group.by = "custom")$data
-  #features.plot has the ensemble ids
-  dotplot_data$name <- annot_subset[dotplot_data$features.plot, "name"]
+  #Construct the dotplot.groups slot
+  for (i in seq_along(group_by_cell_sets)) {
+    cell_set <- group_by_cell_sets[[i]]
+    filtered_cells <- intersect(cell_set$cellIds, cells_id)
+    #This covers a border case where two cell_sets have the same name (but different ID). Can happen in scratchpad
+    if (cell_set$name %in% data$dotplot.groups) {
+      data$dotplot.groups[cells_id %in% filtered_cells] <- paste0(cell_set$name, i)
+    } else {
+      data$dotplot.groups[cells_id %in% filtered_cells] <- cell_set$name
+    }
+  }
+  # If NA values are left in the group, dotplot function will fail.
+  data <- subset(data, subset = dotplot.groups != "NA")
```

Comment:
```suggestion
  data <- subset(data, subset = !is.na(dotplot.groups))
```

untested, but wouldn't typically check for `NA` by testing for equality with the string `"NA"` as it would break if the user defines a dotplot group as "NA"?

Reply:
Oh this was sloppy on my part, you are right